### PR TITLE
Fix Issue 10131 - To remove duplicates and keep order

### PR DESCRIPTION
**Pull request by RazvanN7**
This is the same as #5774 . I closed the PR and made a new one to some github branch mix-ups I made.

### CODE REVIEW

File: std/experimental/algorithm.d [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+///
+@safe unittest
+{
+    import std.algorithm.comparison : equal;
+
+    assert("AAAA".distinct.equal("A"));
+    assert("ABAC".distinct.equal("ABC"));
+    assert([1, 2, 3, 1, 2, 1, 1].distinct.equal([1, 2, 3]));
```
Comment (trikko) [inline std/experimental/algorithm.d]: I think an example with infinite ranges could be useful. Something like:
```d
	auto ascInt = recurrence!((x,n) => n)(0);    				// 0,1,2,3,4,5,6...
	auto descInt = recurrence!((x,n) => n)(0).map!(x => 10-x);  // 10,9,8,7,6,5,4 ..
	
	roundRobin(ascInt, descInt).distinct.take(20).writeln;

```

File: std/experimental/algorithm.d [hunk 2]
```diff
@@ -0,0 +1,49 @@
+// Written in the D programming language.
+
+/**
+This module implements experimental additions/modifications to $(MREF std, _algorithm).
+
+Use this module to test out new functionality for $(REF wrap, std, _algorithm)
```
Comment (wilzbach) [inline std/experimental/algorithm.d]: `wrap`?

File: std/experimental/algorithm.d [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+module std.experimental.algorithm;
+
+import std.range : isInputRange, ElementType;
+
+/**
+Lazily produces a range with no duplicated elements from the input
+range $(D r). The order of elements is the original order.
+
+Params:
+    Range r - the input range to eliminate duplicates from
```
Comment (wilzbach) [inline std/experimental/algorithm.d]: Use `r = input range from which duplicates should be eliminated`

File: std/experimental/algorithm.d [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+Returns:
+    A range containing all the unique elements
```
Comment (wilzbach) [inline std/experimental/algorithm.d]: Nit: misses a final dot ;-)

File: std/experimental/algorithm.d [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+Authors:   Andrea Fontana
+           Razvan Nitu
+ */
+module std.experimental.algorithm;
+
+import std.range : isInputRange, ElementType;
+
+/**
+Lazily produces a range with no duplicated elements from the input
+range $(D r). The order of elements is the original order.
```
Comment (wilzbach) [inline std/experimental/algorithm.d]: Should note that this uses the GC if we really want to go down this road.

File: std/experimental/algorithm.d [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+///
+@safe unittest
+{
+    import std.algorithm.comparison : equal;
+
+    assert("AAAA".distinct.equal("A"));
+    assert("ABAC".distinct.equal("ABC"));
+    assert([1, 2, 3, 1, 2, 1, 1].distinct.equal([1, 2, 3]));
+}
```
Comment (wilzbach) [inline std/experimental/algorithm.d]: This should have a lot more internal test. Please take a lot at other Phobos function for an impression - basic things, for example:
- `const`
- DummyRanges (e.g. RefRanges)
- infinite ranges
- strings (especially unicode ones)
- `nothrow`

File: std/experimental/algorithm.d [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    Range r - the input range to eliminate duplicates from
+
+Returns:
+    A range containing all the unique elements
+*/
+auto distinct(Range)(Range r)
+if (isInputRange!Range)
+{
+    import std.algorithm.iteration : filter;
+    bool[ElementType!Range] justSeen;
```
Comment (wilzbach) [inline std/experimental/algorithm.d]: When I suggested moving this to `stdx.algorithm`, I was suggesting that we can use this to experiment with APIs that allow the user to control how the allocation happens.
There are a couple of ways to do this. Some quick ideas (there are more!):
- Do the deallocation automatically, e.g. like RefCounted and provide an optional Allocator
- Let the user do the deallocation (e.g. by returning the range and the allocated array as tuple)

Please note that it's very important that this function can be used in `@nogc`.

**Comment by andralex** [comment]
I'll add that this (and other related functions) is an important function for things like relational joins etc. Martin Odersky showed me some pretty amazing short example that do powerful things because of such.

**Comment by andralex** [comment]
After planning things with @RazvanN7 will close this in wait for scoped hashtables. cc @edi33416 :)

**Pull request closed without merging**